clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -1629,37 +1630,79 @@ void FunctionVariableConstraint::brainTransplant(ConstraintVariable *FromCV,
   auto RetVar = getOnly(returnVars);
   RetVar->brainTransplant(FromRetVar, I);
   // Transplant params.
-  assert(From->numParams() == numParams());
-  for (unsigned i = 0; i < From->numParams(); i++) {
-    CVarSet &FromP = From->getParamVar(i);
-    CVarSet &P = getParamVar(i);
-    auto FromVar = getOnly(FromP);
-    auto Var = getOnly(P);
-    Var->brainTransplant(FromVar, I);
+  if (numParams() == From->numParams()) {
+    for (unsigned i = 0; i < From->numParams(); i++) {
+      CVarSet &FromP = From->getParamVar(i);
+      CVarSet &P = getParamVar(i);
+      auto FromVar = getOnly(FromP);
+      auto Var = getOnly(P);
+      Var->brainTransplant(FromVar, I);
+    }
+  } else if (numParams() != 0 && From->numParams() == 0) {
+    auto &CS = I.getConstraints();
+    vector<ParamDeferment> &defers = From->getDeferredParams();
+    assert(getDeferredParams().size() == 0);
+    for(auto deferred : defers ) {
+      assert(numParams() == deferred.PS.size());
+      for(unsigned i = 0; i < deferred.PS.size(); i++) {
+        CVarSet ParamDC = getParamVar(i);
+        CVarSet ArgDC = deferred.PS[i];
+        constrainConsVarGeq(ParamDC, ArgDC, CS, &(deferred.PL), Wild_to_Safe, false, &I);
+      }
+    }
+  } else {
+    llvm_unreachable("Brain transplanting to function w/ empty param vars");
   }
 }
 
-void FunctionVariableConstraint::mergeDeclaration(ConstraintVariable *FromCV) {
+void FunctionVariableConstraint::mergeDeclaration(ConstraintVariable *FromCV,
+                                                  ProgramInfo &I) {
+  // `this`: is the declaration the tool saw first
+  // `FromCV`: is the declaration seen second, it cannot have defered constraints
   FVConstraint *From = dyn_cast<FVConstraint>(FromCV);
   assert (From != nullptr);
+  assert(From->getDeferredParams().size() == 0);
   // Transplant returns.
   auto FromRetVar = getOnly(From->getReturnVars());
   auto RetVar = getOnly(returnVars);
-  RetVar->mergeDeclaration(FromRetVar);
+  RetVar->mergeDeclaration(FromRetVar, I);
+  bool isEmpty = numParams() == 0;
   // Transplant params.
-  assert(From->numParams() == numParams());
-  for (unsigned i = 0; i < From->numParams(); i++) {
-    CVarSet &FromP = From->getParamVar(i);
-    CVarSet &P = getParamVar(i);
-    auto FromVar = getOnly(FromP);
-    auto Var = getOnly(P);
-    Var->mergeDeclaration(FromVar);
+
+  bool dealingWithUntyped = isEmpty && numParams() != From->numParams();
+  auto &CS = I.getConstraints();
+
+  if(From->numParams() == 0) {
+    // From is an untyped declaration, and adds no information
+    return;
+  } else if (this->numParams() == 0) {
+    // This is an untyped declaration, we need to perform a transplant
+    From->brainTransplant(this, I);
```

Comment:
I think this might not work. The problem is that while the call to `brainTransplant` will generate the deferred constraints from `this` on the variables in `From`, I think we we might need to update the `Map` in `ProgramInfo::insertIntoExternalFunctionMap` to map to `From` rather than `this`.

That is making me think that we have to update that code.

Maybe change line 494 to be
```
bool isDef = newC->hasBody() || (oldC->numParams() == 0 && newC->numParams() != 0);
```

Reply:
If we make this change, then we should know that `this->numParams()  != 0` in this code unless `From->numParams() == 0` so we can just delete this `else if` part.

Reply:
Probably the `oldC->numParams()` check also needs to know if this was a legit no-arg function or a non-prototype.

Reply:
Actually, I think this could work:
```
bool isDef = newC->hasBody() || (!oldC->hasBody() && oldC->numParams() == 0 && newC->numParams() != 0);
```